### PR TITLE
Add pre-commit with black, ruff and pylint automation

### PR DESCRIPTION
**Pull request by qartik**
Also:
- improve README
- update package versions
- add Makefile
- resolve typos and linting issues reported by above tools
- remove lark warnings from pytest output

### CODE REVIEW

File: .github/workflows/docs/build-docs [hunk 1]
```diff
@@ -28,7 +28,7 @@ def remove_dir(dirpath):
 
 
 def fix_links(filepath):
-    with open(filepath, "r", encoding="utf8") as f:
```
Comment (cqc-melf) [inline .github/workflows/docs/build-docs:31]: I am still not sure why this change is made, have you done this manually or is ruff doing this changes? I have done a PR with only updating black and that is not doing this change.

File: .github/workflows/docs/conf.py [hunk 2]
```diff
@@ -1,5 +1,3 @@
-# -*- coding: utf-8 -*-
```
Comment (cqc-melf) [inline .github/workflows/docs/conf.py:1]: See above

File: .github/workflows/docs/conf.py [hunk 3]
```diff
@@ -92,7 +91,8 @@
 def add_reference(
     app: Sphinx, env: BuildEnvironment, node: Element, contnode: TextElement
 ) -> Optional[nodes.reference]:
-    # Fix references in docstrings that are inherited from the base pytket.backends.Backend class.
+    # Fix references in docstrings that are inherited
```
Comment (cqc-melf) [inline .github/workflows/docs/conf.py:94]: See above

File: .github/workflows/docs/build-docs (hunk 1 above)
Comment (qartik) [reply .github/workflows/docs/build-docs:31]: Same answer as before: https://github.com/CQCL/pytket-qir/pull/57#discussion_r1280597356 Ruff does it with its implementation of the pyupgrade linter.

Copying over:

> It's the default option and [hence unneeded](https://beta.ruff.rs/docs/rules/redundant-open-modes/).

File: .github/workflows/docs/conf.py (hunk 2 above)
Comment (qartik) [reply .github/workflows/docs/conf.py:1]: See https://github.com/CQCL/pytket-qir/pull/57#discussion_r1280599374 and later comments in the thread.

File: .github/workflows/docs/conf.py (hunk 3 above)
Comment (qartik) [reply .github/workflows/docs/conf.py:94]: This change is manual to conform to the line width limit.

File: .github/workflows/docs/build-docs (hunk 1 above)
Comment (cqc-melf) [reply .github/workflows/docs/build-docs:31]: I am fine with doing this change, maybe my question was not that clear. Had ruff done this automatically or have you done this change manually? And if ruff has done that, with which command run?

File: README.md [hunk 4]
```diff
@@ -32,12 +30,32 @@ There is also a Slack channel for discussion and support. Click [here](https://t
 
 ## Development
 
-To install this extension in editable mode, simply change to this directory, and run:
+First setup your virtual environment and install required dependencies:
+
+```sh
+python -m venv .venv
```
Comment (cqc-melf) [inline README.md:36]: See https://github.com/CQCL/pytket-qir/pull/57#discussion_r1280635199

I think it would be good to do something like

```
First install required dependencies:

```sh
pip install -U pip setuptools
pip install build pre-commit pytest wheel mypy~=1.4 black~=23.7 pylint~=2.17 ruff==0.0.282
pre-commit install
```

Then install this extension in editable mode, simply change to this directory, and run:

File: .github/workflows/docs/build-docs (hunk 1 above)
Comment (qartik) [reply .github/workflows/docs/build-docs:31]: `ruff check .` shows all the lint violations. Some of those can be automatically corrected by ruff using `ruff check . --fix`. But in this PR, we also have `pre-commit` set up in such a way that `ruff check --fix` will be automatically called on all files that are being added/modified in the commit. The last sentence is True only if `pre-commit install` has been issued once in your local clone once as mentioned in the README in the development steps. I am happy to show a demo if desired.

File: README.md (hunk 4 above)
Comment (qartik) [reply README.md:36]: What's the argument against providing steps for creating a virtual environment for beginners? I like to provide a self-contained set of steps (especially those that I can copy and paste myself without having to think). People who are already familiar with virtual environments and prefer setting them up any other way can just skip copying the first two lines, no?

**Review by cqc-melf — APPROVED** [review]
